pmlc/conversion/gpu/gather_gpu_launch_funcs.cc

```diff
@@ -0,0 +1,82 @@
+// Copyright 2020 Intel Corporation
+
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+
+#include "pmlc/conversion/gpu/pass_detail.h"
+
+using namespace mlir; // NOLINT[build/namespaces]
+
+namespace pmlc::conversion::gpu {
+namespace gpu = mlir::gpu;
+
+struct Move {
+  mlir::Operation *insertionOp;
+  std::vector<mlir::Operation *> ops;
+};
+
+class GatherGpuLaunchFuncsPass
+    : public GatherGpuLaunchFuncsPassBase<GatherGpuLaunchFuncsPass> {
+public:
+  void runOnOperation() override;
+
+private:
+  mlir::Operation *
+  gatherAllocOpsStartingFromOp(mlir::Operation *op,
+                               std::vector<mlir::Operation *> &ops);
+  void gatherAllocOpsBetweenConsecutiveLaunchOps(
+      mlir::Block &block, std::vector<std::shared_ptr<Move>> &moves);
+};
+
+mlir::Operation *GatherGpuLaunchFuncsPass::gatherAllocOpsStartingFromOp(
+    mlir::Operation *op, std::vector<mlir::Operation *> &ops) {
+  while (op != nullptr) {
+    if (mlir::isa<mlir::AllocOp>(op)) {
+      ops.push_back(op);
+    }
+    if (!mlir::isa<mlir::AllocOp>(op) && !mlir::isa<gpu::LaunchFuncOp>(op)) {
+      return op;
+    }
+    op = op->getNextNode();
+  }
+  return op;
+}
+
+void GatherGpuLaunchFuncsPass::gatherAllocOpsBetweenConsecutiveLaunchOps(
+    mlir::Block &block, std::vector<std::shared_ptr<Move>> &moves) {
+  mlir::Operation *op = &(*block.begin());
+  while (op != nullptr) {
+    if (mlir::isa<gpu::LaunchFuncOp>(op)) {
+      auto move = std::make_shared<Move>();
+      move->insertionOp = op;
+      op = gatherAllocOpsStartingFromOp(op, move->ops);
+      moves.push_back(move);
+    }
+    if (op != nullptr) {
+      op = op->getNextNode();
+    }
+  }
+}
+
+void GatherGpuLaunchFuncsPass::runOnOperation() {
+  auto module = getOperation();
+  module.walk([&](mlir::FuncOp funcOp) {
+    if (funcOp.isExternal()) {
+      return;
+    }
+    auto &blocks = funcOp.getBlocks();
+    for (auto &block : blocks) {
+      std::vector<std::shared_ptr<Move>> moves;
```

Comment:
Nitpick, but why not return the vector of Moves rather than pass by reference?

Reply:
Changed in [PR1605](https://github.com/plaidml/plaidml/pull/1605/files#diff-df1e828d87ae2c07b4b3674b23ed4c4e560635639564f3acb37d3bcdf56ca021R27)